Unified build

Biont opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added/updated (for bug fixes/features)

**What kind of change does this PR introduce?** 
Feature - New reusable workflow for building and pushing plugin/theme archives to build branches

**What is the current behavior?**
Previously, we had a workflow for creating plugin archives that required manual artifact storage and management. The build process was more complex and didn't provide an automated way to access builds directly via GitHub branches.

**What is the new behavior (if this is a feature change)?**

This PR introduces a new `build-push.yml` reusable workflow that:

### Core Functionality
- **Automatic build branch determination**: Inspects the origin branch and creates corresponding build branches while respecting our branching strategy  `dev/` prefix (e.g., `dev/main` → `main`)
- **Intelligent version handling**: When `PACKAGE_VERSION` is not provided, automatically fetches the latest release and creates semver-compatible pre-release versions using the normalized branch name (e.g., `1.2.3-feature-user-auth`)
- **Comprehensive build process**: Handles both plugins and themes with proper version header updates, asset compilation, and dependency management

### Build Process Features
- Installs production dependencies via Composer with configurable arguments
- Compiles assets using `npm run build` with Node.js setup
- Updates version information in plugin/theme headers and `package.json`
- Integrates with WordPress Translation Downloader when configured
- Supports PHP-Scoper for dependency prefixing with proper autoloader handling
- Applies `.distignore` file filtering to exclude development files
- Commits and pushes build artifacts to determined build branches
- Creates GitHub Actions artifacts for download (excluding `.git` folder)

### Key Benefits
- **Direct GitHub access**: Build artifacts are accessible directly via build branches, eliminating need for external artifact storage
- **Automatic versioning**: Reduces manual version management overhead while maintaining semver compliance
- **Flexible configuration**: Supports extensive customization through inputs and secrets
- **Signed commits**: Supports SSH key-based commit signing for security
- **Clean artifacts**: Properly excludes `.git` folder and respects `.distignore` patterns

### Configuration Options
- Configurable Node.js and PHP versions
- Support for private npm registries
- Custom PHP tools installation
- Flexible Composer arguments
- Pre-script execution for custom build steps
- Override capability for build branch naming
- Comprehensive environment variable support

**Does this PR introduce a breaking change?**
No breaking changes to existing workflows. This is a new workflow that can be adopted independently. However, teams adopting this workflow should be aware of:

- **New branch naming convention**: Build artifacts will be pushed to `dev/{origin-branch}` branches
- **Version format changes**: Auto-generated versions follow `{latest-release}-{normalized-branch}` pattern
- **Required secrets**: Workflow requires specific GitHub user credentials for commit signing and pushing

**Other information**:

### Migration Path
Teams can gradually migrate from existing build processes by:
1. Adding the new workflow alongside existing ones
2. Testing with non-critical branches first
3. Updating deployment processes to reference new build branches
4. Removing old workflows once confidence is established

### Security Considerations
- All commits are signed using SSH keys for authenticity
- Supports private npm registries with token authentication
- Composer authentication for private packages
- Proper cleanup of temporary signing keys

### Documentation
- Comprehensive documentation added following existing patterns
- Examples for both simple and complex use cases
- Clear explanation of automatic version handling
- Best practices and recommendations included

### Testing Recommendations
- Test with different branch naming patterns
- Verify version generation with and without existing releases
- Confirm build artifacts exclude appropriate files
- Validate signed commits and build branch creation

This workflow significantly improves our CI/CD capabilities by providing a standardized, automated approach to building and distributing plugin/theme archives while maintaining flexibility for project-specific requirements.

tyrann0us committed:
WIP

Commit:
WIP

Biont committed:
🎨 Update version in `package.json` during build process

This change ensures that the plugin version is updated in `package.json` without committing a new git tag, as specified by the `--no-git-tag-version` flag.

Biont committed:
🛠 Add artifact upload step to GitHub Actions workflow

This commit adds a new step to the build-push.yml workflow to upload artifacts. The artifact name is dynamically set based on the package folder name and plugin version, with default values provided for each.

Biont committed:
🎯 Add Composer auth support and update PHP tool setup

This commit adds support for authenticating with privately hosted packages using `COMPOSER_AUTH_JSON` secret. It also updates the PHP tools setup to include a step for installing Composer dependencies without dev dependencies, allowing more flexible configuration options for CI workflows.

Biont committed:
🔄 Only output matching filenames in grep calls

Biont committed:
🔧 Fix conditional logic in theme detection

Biont committed:
🔧 Set up signing commity similar to other workflows

Biont committed:
🔧 Introduce BUILT_BRANCH_NAME

Biont committed:
🔧 Improve distignore handling

Biont committed:
🔧 Improve distignore handling

Biont committed:
🔧 checkout ref branch onto build branch

Biont committed:
🔧 Improve & fix variable usage

Biont committed:
🔧 Force unstage

Biont committed:
🔧 Fix package name fallback

Biont committed:
Improve branch handling

- Added logic to determine the build branch name dynamically based on input or origin branch.
- Modified commit message to include the target branch for clarity.
- Simplified asset addition process by using `git add -A`.
- Ensured that push operation is conditional, only executing if changes are detected.
- Added steps to delete temporary signing key files.

Biont committed:
Optionally use latest release version as package version

Ensure optional handling of `PACKAGE_VERSION`. Fetch latest release if not provided, appending normalized branch name as pre-release identifier. Adjust workflow steps accordingly. 📘 This update enhances the flexibility and automation of the package versioning process in GitHub Actions workflows.

Biont committed:
Use correct repo name and exclude `.git` in artifact upload

Ensure the workflow uses the correct repository name for package folder naming and excludes the `.git` directory when uploading artifacts to prevent unnecessary files from being included.

Biont committed:
Add documentation for the Build and Push reusable workflow

Biont committed:
Rename job name and workflow file reference to `build-push`

Biont committed:
Add blank lines as suggested by linter

Biont committed:
Merge branch 'main' into unified-build

Chrico commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,289 @@
+          fi
+          
+          echo "PACKAGE_VERSION=$PACKAGE_VERSION" >> $GITHUB_ENV
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_USER_SSH_KEY && secrets.GITHUB_TOKEN || github.token }}
```
can we put the `env` before `run` ?

Chrico commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,289 @@
+
+      - name: Determine package version
+        run: |
+          if [ -n "${{ inputs.PACKAGE_VERSION }}" ]; then
+            PACKAGE_VERSION="${{ inputs.PACKAGE_VERSION }}"
+            echo "Using provided package version: $PACKAGE_VERSION"
```
Is this "if" required? The `env.PACKAGE_VERSION` is already set above the steps. So maybe we can run this step with an `if:`-condition?

Chrico commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,289 @@
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Set plugin version header
+        if: ${{ hashFiles('style.css') == '' }}
```
`style.css` is part of WordPress Themes. Do we need to adjust the `if:`-condition here to `index.php` or do we need to update the `name:` to "Theme"? 👀

Biont replied:
Yeah we've been tripping over this before. I think we can afford a variable that clearly tells us what this is about

Chrico commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,289 @@
+        run: |
+          sed -Ei "s/Version: .*/Version: ${{ env.PACKAGE_VERSION }}/g" $(grep -rl --include "*.php" "Plugin Name:" . | head -1)
+
+      - name: Add commit hash to plugin header
+        if: ${{ hashFiles('style.css') == '' }}
```
Same question as https://github.com/inpsyde/reusable-workflows/pull/194/files#r2194270750

Biont committed:
🤖 Update build-push documentation and workflow for asset compilation 🚀

This commit updates the `build-push.md` documentation and `.github/workflows/build-push.yml` file. It implements an inverted branching scheme where development branches are prefixed with `dev/`, and build branches contain compiled assets without the `dev/` prefix. The version handling process has been updated to strip the `dev/` prefix from branch names and normalize them for semver compatibility. Additionally, migration steps are provided in the documentation to help users transition from previous workflows.

Biont committed:
Merge remote-tracking branch 'origin/unified-build' into unified-build

Chrico commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,289 @@
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Set plugin version header
```
I'm adding this to the top of the whole section. Shoudl we maybe in general combine "Prepare Plugin" and "Prepare Theme" as 2 steps in total instead of having multiple steps and checks for version, commit hash, ...?

Biont committed:
🔧 Handle WordPress themes and plugins uniformly.

Biont commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,289 @@
+      # During development, symlinking is preferable.
+      # In production builds, we ship the actual install location and remove the source directory.
+      COMPOSER_MIRROR_PATH_REPOS: 1
+      GIT_SHA: ${{ github.sha }}
```
Required not. It could theoretically be cleaner to access in shell scripts, but we are not even doing that (still using `${{ ... }}` syntax ), so I guess we can change this

Biont committed:
🔧 Remove unused `GIT_SHA` environment variable and update references

Biont committed:
🔧 Put "env:" before "run:" for better visibility

Biont committed:
🌟 Streamline `Determine package version` step in GitHub Actions workflow

Revised the logic to remove redundant checks and simplify the environment variable assignment by integrating conditionals directly into the flow.

Biont commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,289 @@
+          tools: php-scoper, sniccowp/php-scoper-wordpress-excludes, composer/installers, inpsyde/wp-translation-downloader, ${{ inputs.PHP_TOOLS }}
+          coverage: none
+
+      - name: Install Composer dependencies without dev dependencies
```
This is unchanged from the existing `build-plugin-archive.yml` workflow. I actually believe there even was a time when the `--no-dev` was hardcoded next to other args coming from the input.

But this also harkens back to a time when we needed to check out composer twice:
Once for getting all the tooling (asset compiler, translation downloader etc) and once for getting the "production-ready" state. This is no longer needed. 

So today, I think this is mostly about the wording. There isn't actually a need to spell out the "without dev dependencies" here any longer, so I'm just going to remove it. The input/config is fine in my book, but I am also not married to it. For now I'll just change the comment because it is the smaller change. Let me know if this works for you.

Biont committed:
📝 Update step description to remove noise

Biont committed:
📝 Update list formatting

Biont committed:
🎉 Add validation to ensure source and build branches are separate

This commit introduces a step in the GitHub Actions workflow to validate that the branch where build artifacts will be committed is different from the branch containing the source code. This prevents accidental commits of build artifacts back into the development branches, enforcing a clean separation between source code and built outputs.

The validation checks if the `BUILT_BRANCH_NAME` environment variable matches the current branch (`ORIGIN_BRANCH`). If they match, it prints an error message explaining the potential issue and exits with a status of 1 to halt the workflow. This ensures that builds are always pushed to designated release branches rather than inadvertently back into feature or development branches.

Additionally, the commit provides guidance on how to properly configure the `BUILT_BRANCH_NAME` using either branch naming conventions (e.g., prefixing dev branches) or by explicitly setting the input in the workflow trigger.

tyrann0us reviewed (commented):
Thank you so much for working on this! This could be a real game-changer and massively improve cross-team collaboration with a unified asset compilation. No more learning about other workflows or strange variations. Thank you for that! 💪🏽
I finally took the time to review it because it will be OOF for the next two weeks, and I didn't want to block this any longer. I left several mostly minor comments and suggestions.

Biont committed:
🚀 Tidy up input parameters and descriptions

Biont commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,314 @@
+        default: '8.2'
+        required: false
+        type: string
+      PACKAGE_FOLDER_NAME:
+        description: The name of the package folder (falls back to the repository name).
+        required: false
+        default: ''
+        type: string
```
Okay this is a bit unfortunate. There are several angles here

- **Some plugins need to be able to customize the plugin folder** so that the _install location_ can be chosen independently from the repo/archive name. Usually this has to do with maintaining compatibility with an existing plugin slug on wp-org SVN - and maintaining a frictionless (auto-)update path for existing plugin users. But scenarios are also imaginable where you need to stay consistent with an existing WordPress installation despite extracting a bunch of plugins into a `<CLIENT>-awesome-plugin` repo naming scheme

- **We've been asserting that plugin archives do not need to contain a subfolder to be installable**. The existing workflow (or rather: the  **wp dist-archive** command ) [does this](https://github.com/inpsyde/reusable-workflows/blob/a810bb31acd37bfbf209c5da79e65b9ade003495/.github/workflows/build-plugin-archive.yml#L279) - and here we've previously been placing the folder customization. But since it's not a technical requirement and not commonly used in agency projects that use build branches, we decided to approach this with a KISS philosophy.

- **Build branches with only a folder in it look pretty bad**. Yes this is purely aesthetics and I do not expect developers to often browse these branches, but I remember us discussing that it looks less confusing without.

- We frequently hand over individual feature packages to QA. So a way to make sense of individual archives - as well as maintaining a consistent install folder name - could be a requirement.
---

The new workflow - as of now - would result in an install folder like this if you install from a zip file (either from artifact or manual GutHub zip download):

<img width="991" height="609" alt="screenshot_2025-07-29-090009" src="https://github.com/user-attachments/assets/ee6dd4ef-7025-4524-b817-5b83abc5ca6f" />

This is because we combine `PACKAGE_FOLDER_NAME` and `PACKAGE_VERSION`. For manual source code downloads, it combines the repo name with the current tree/branch. 

Whereas when I put it all into a dedicated subfolder, it looks like this:
<img width="1123" height="742" alt="screenshot_2025-07-29-090938" src="https://github.com/user-attachments/assets/387b45c8-6ecf-41d5-a54b-efef79d27c2e" />

I guess WP falls back to the zip filename if no subfolder is present.

---

So all things considered I feel there is no way to avoid a subfolder inside the build branches. Are there any experiences with the composer installer workflows that would speak against this? That's the one thing I did not check

Biont replied:
https://github.com/inpsyde/reusable-workflows/pull/194/commits/5f04664ecf828d015f648a845a3e5918ec2abfe1 introduces a subfolder within archives

Biont committed:
🌍 Rename and update `NO_CHANGES` environment variable in GitHub workflow

Biont committed:
🚀 Tidy up input parameters and descriptions

Biont committed:
🚀 Remove whitespace

Biont committed:
✨ Improve environment variable comments in GitHub workflow

Biont commented on docs/build-push.md:
```diff
@@ -0,0 +1,255 @@
+
+## Branch naming convention
+
+The workflow implements an inverted branch scheme where development branches are prefixed with `dev/` and build branches contain the compiled assets:
```
Because it reserves the "top-level" branch segment for release artifacts and thereby turns the actual source code into a second-class citizen `main (artifact) <-- dev/main (source)`.

All popular git flows I'm aware of would keep source code at the top, then use `feature/foo-bar` and `release/1.2.3` as a convention for separate purposes.

It would follow to use something like `build/1.2.3` for release artifacts if we're looking at this purely through the lens of git branching strategies. 

If I understood correctly, the setup we chose is mainly because it's more convenient for composer - ie. `composer require syde/acme:dev-main` instead of `composer require syde/acme:dev-build/main`.

Biont committed:
✨ Tidy up docs and output

Biont committed:
🎉 Switch release listing method from `gh` to `git`

Use the built-in Git functionality for fetching the latest release instead of relying on the GitHub CLI tool (`gh`). This makes it work for repositories that rely solely on tags and do not use GH releases

Biont commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,341 @@
+          GITHUB_TOKEN: ${{ secrets.GITHUB_USER_SSH_KEY && secrets.GITHUB_TOKEN || github.token }}
+        run: |
+          # Fetch latest public release
+          LATEST_RELEASE=$(gh release list --limit 1 --exclude-drafts --exclude-pre-releases --json tagName --jq '.[0].tagName' 2>/dev/null || echo "")
```
I researched an alternative approach that 

1. Works with git instead of gh
2. Does not rely on releases - works purely with tags
3. Is quite a bit faster (I assume the gh cli does some API calls here)

The theoretical downside is that **tags** are not **releases**: You can create tags for completely arbitrary reasons ( `before-giant-refactor`, `last-php7-compatible` etc - purely in order to label commits as human-friendly points in time you may wish to jump back to ). A GH Release -while being backed by tags carries at least some credible expectation of being tied to a (semantic) version number.

So this workflow might run into problems in repos that use both sem-ver releases as well as arbitrary tags: It would simply pick the latest tag instead of the latest release in these cases and yield undesired results.

But I guess this is an edge case - and we can still use `inputs.PACKAGE_VERSION` in the calling workflows to specify the latest release via the gh cli.

Biont committed:
📝 Update build script to use `wp-scripts` for consistency.

The change updates the build script command in the documentation from using Webpack directly to utilizing WordPress's built-in wp-scripts.

Biont committed:
✨ Refactor workflow documentation for clarity and best practices

Biont committed:
🌱 Update heading level and format in build-push.md

Biont committed:
✨ Simplify build and push documentation

This commit consolidates the description of version updates for both plugins and themes

Biont committed:
🚀 Streamline GitHub token usage in build workflow

Chrico commented on .github/workflows/build-push.yml:
In this case, when outputting data with `<type>: <value>`, no full stop is required. But in case of full sentences, we should add a full stop.

Example: https://github.com/inpsyde/reusable-workflows/blob/de379f96149671a6196f16f469ad77df5b3431ed/.github/workflows/build-push.yml#L140

> No releases found, using default version 0.0.0.

Biont replied:
https://github.com/inpsyde/reusable-workflows/pull/194/commits/6eace1721b425a73b9c164b83e6caebbce8a4f9f

Chrico commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,335 @@
+            echo "Package type: WordPress Theme"
+            echo "IS_THEME=true" >> $GITHUB_ENV
+          else
+            echo "Package type: WordPress Plugin"
```
What about Libraries?

Chrico replied:
I suppose it's just a matter of wording. :) We have a few libraries that also ship JS/CSS and need the build workflow. 💪🏻

Biont replied:
I refactored things a little, so that we are only checking against a generic `PACKAGE_TYPE` from now on. This would make it easier to address types other than plugins/themes in the future.
https://github.com/inpsyde/reusable-workflows/pull/194/commits/d0deb15ae4ec6a34a6afd8cc847d958960ae1ee2

Chrico commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,341 @@
+        if: ${{ env.IS_THEME == 'false' }}
+        run: |
+          # Find and update plugin main file
+          PLUGIN_FILE=$(grep -rl --include "*.php" "Plugin Name:" . | head -1)
```
> I am certain we did not. Not only would that prevent existing plugin projects from adopting the workflow, it is simply not a convention that exists outside of our company. The official developer guidelines explicitly ask for the file to be named after the containing folder.

We should align here. While we state in a few areas "this is how you have to use this workflow because _we_ decided it", here we go the other way around and state "outside of our company".

Biont replied:
https://github.com/inpsyde/reusable-workflows/pull/194/commits/044964f1bc1b8ca8901fb0a93d8140837b933442

Chrico commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,335 @@
+            echo "Warning: No plugin main file found!"
+          fi
+
+      - name: Prepare Theme
```
While I think it is good to have a check for the Plugins Headers and Theme style.css Headers, should we maybe also support something dynamic like a global constant, which could also contain the version? This could become handy for libraries that do not have a version set anywhere in code and are parseable through WordPress.

Chrico replied:
We have some libraries (not Plugins) which also contain a version that is shown in the Backend, or, depending on the version, something needs to be triggered (like updating the DB tables). In this case, a version string also needs to be added and updated manually somewhere in a file. 
Through `inpsyde/modularity`, we only have the possibility right now to set `version` in the `composer.json` for libraries - see https://github.com/inpsyde/modularity/blob/master/src/Properties/LibraryProperties.php#L89-L93

My thought here was to also cover the need to have a "version" changed in code for "PHP libraries".

Biont replied:
https://github.com/inpsyde/reusable-workflows/pull/194/commits/91021336700af3e9011e019bf2181dd883d2289b

Biont committed:
🛠️ Update workflow to update version in `composer.json`

This commit adds a step to the GitHub Actions workflow to update the version in the `composer.json` file, ensuring consistency with the version set in `package.json`. This is necessary because our application relies on both npm and composer for dependency management.

Biont committed:
✨ Add package type determination for themes and plugins

This commit enhances the GitHub Actions workflow by adding logic to determine whether a package is a WordPress theme, plugin, or generic library. The changes include:
- Enhancing the `Determine package type` step to check for specific headers in `style.css` for themes and PHP files for plugins.
- Introducing environment variable `PACKAGE_TYPE` to store the determined package type.
- Modifying subsequent steps (`Prepare WordPress Plugin`, `Prepare WordPress Theme`) to conditionally run based on the new `PACKAGE_TYPE`.
- Adding a fallback step for generic libraries when no theme or plugin headers are found.

Biont committed:
🔄 Rename PACKAGE_FOLDER_NAME to PACKAGE_NAME in .github/workflows/build-push.yml

This commit renames the input parameter and environment variable from `PACKAGE_FOLDER_NAME` to `PACKAGE_NAME`, which more accurately reflects the content of the package rather than its folder structure. The changes also update the artifact preparation step to ensure that the compiled package is copied into a subdirectory with the name specified by `PACKAGE_NAME`.

Biont committed:
📝 Remove redundant passage

Biont committed:
📝 Simplify version and build branch handling in GitHub workflow

Refactored the script to simplify the logic for stripping the 'dev/' prefix from branch names when determining the version string and build branch. This change ensures consistent behavior regardless of whether the branch name starts with 'dev/'. The code is now more concise and easier to understand.

Biont commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,341 @@
+            VERSION_BRANCH="${ORIGIN_BRANCH#dev/}"
+          else
+            VERSION_BRANCH="$ORIGIN_BRANCH"
+          fi
```
https://github.com/inpsyde/reusable-workflows/pull/194/commits/5abc920b4ac5d596fbe974f96b7109c496bdcb7e

Biont committed:
📝 Fix wording

Biont committed:
✅ Enhance and error-proof plugin main file detection in build workflow

This commit updates the `build-push.yml` GitHub Actions workflow to improve the detection of the WordPress plugin's main PHP file. The changes prioritize checking specific files (`index.php` and `${{ env.PACKAGE_NAME }}.php`) for a "Plugin Name:" header. If neither file is found, it outputs an error message with detailed instructions instead of a generic warning, ensuring that users know exactly what needs to be corrected.

Biont committed:
🔄 Updated punctuation

Biont committed:
📝 Update build-push documentation to reflect latest code changes

The workflow now fetches the latest tag from the repository instead of the latest release. This ensures that the versioning is based on Git tags, which aligns better with semantic versioning practices. The commit also renames `PACKAGE_FOLDER_NAME` to `PACKAGE_NAME` and clarifies the purpose of `PRE_SCRIPT`.

Biont committed:
📝 Append commit SHA to package version for better traceability

This change appends the GitHub commit SHA to the generated package version. This provides more context about which specific code changes are included in each build, enhancing traceability and debugging capabilities. By incorporating the commit SHA into the semver format, it ensures that every build is uniquely identifiable.

Biont committed:
🤖 Remove `--delete` option in build-push workflow to prevent accidental deletion of files

The `rsync -av --delete ./build/ . && rm -rf ./build/` command was modified to remove the `--delete` option. We have a dedicated .distignore pass for this concern and this flag actually resulted in the loss of .git and .distignore dirs/files.

Biont committed:
✅ Include short commit SHA in package version for better traceability.

- Updated the `.github/workflows/build-push.yml` file to include the first 7 characters of the commit SHA (`SHORT_SHA`) in the generated package version.
- This change enhances traceability by providing a more granular identifier for builds, especially useful for debugging and tracking down specific commits associated with particular versions.

Chrico commented on .github/workflows/build-push.yml:
```diff
@@ -0,0 +1,366 @@
+  build-push:
+    timeout-minutes: 10
+    runs-on: ubuntu-latest
+    env: # Used to check whether there are changes to commit
```
Just a small note, but can we keep comments consistently always above the line it is affecting? 

```yml
env: 
	# Used to check whether there are changes to commit
	HAS_GIT_CHANGES: ''
	# Disables symlinking of local "path" repositories during package installation.
	# During development, symlinking is preferable.
	# In production builds, we ship the actual install location and remove the source directory.
	COMPOSER_MIRROR_PATH_REPOS: 1
	# Will be overridden by determine package version step
	PACKAGE_VERSION: ${{ inputs.PACKAGE_VERSION }}
	PACKAGE_NAME: ${{ inputs.PACKAGE_NAME && inputs.PACKAGE_NAME || github.event.repository.name }}
```

tyrann0us commented:
Deprecation warning needs to be added to the "old" workflow and documentation.

tyrann0us reviewed (commented):
Thank you for implementing all the changes! Left a final round of very minor comments.

Apart from that, could you please add a deprecation warning to the "old" "build and push" workflow and documentation? Thank you!

Biont committed:
:construction_worker: build(CI): Reformat env block in build‑push workflow to improve readability and comment placement.

Biont committed:
:green_heart: ci(workflow): detect WordPress plugin main file automatically with first PHP file containing Plugin Name header and use it in update step.

Biont committed:
Update .github/workflows/build-push.yml

Co-authored-by: Philipp Bammes <8144115+tyrann0us@users.noreply.github.com>
Signed-off-by: Moritz Meißelbach <arbelzapf@gmail.com>

Biont committed:
:construction_worker: refactor: Clean build branch before commit

Biont committed:
Add previously untracked/ignored files

Biont committed:
Specifically only add files not specified in .distignore

Biont committed:
Revert ":construction_worker: refactor: Clean build branch before commit"

This reverts commit e98dc4886f58d5b74d883202f800437887d457fc.

Biont committed:
:recycle: refactor(build-push.yml): Clean up all the remaining .gitignore files in the repository.

Biont committed:
Reapply ":construction_worker: refactor: Clean build branch before commit"

This reverts commit 533102d1aba7aa7c26fb62414a16606efe8396c1.

Biont committed:
:arrow_up: refactor(build-push): Restore previous behaviour when refreshing the working tree

tyrann0us committed:
chore: add deprecation warnings

tyrann0us committed:
fix: fix YAML typos

tyrann0us committed:
docs: add deprecation warning

tyrann0us committed:
fix: stray character

tyrann0us reviewed (approved):
> Apart from that, could you please add a deprecation warning to the "old" "build and push" workflow and documentation? Thank you!

I added the deprecation warnings. @Chrico FYI, this now officially deprecates three workflows:
1. "Build and push assets"
2. "Assets compilation" (with Composer Assets Compiler)
3. "Create plugin archive"

Are you fine with that?

Biont committed:
:arrow_up: refactor(build-push): Improve README

Chrico committed:
Update docs/build-push.md

Co-authored-by: Erika Gili <e.gili@syde.com>
Signed-off-by: Christian Leucht <family.leucht@gmail.com>

Chrico commented on docs/build-push.md:
```diff
@@ -0,0 +1,216 @@
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
```
Hard to say. If we have 2 PR's merged, only the last one that was merged matters, because this one also contains the changes from the first, right? The problem I see here is that we have merged the following:

- PR 1 -> merge
- PR 2 -> merge after PR 1 and has changes from 1 because 1 is "latest state" in "dev/master".

But if, for whatever reason, the build process for PR 1 takes longer than PR 2, then we will have as a result in the built repo "PR 1" pushed. 👀

Biont committed:
:recycle: refactor(workflows): Rename build-push workflows to build-and-distribute and update references

Biont committed:
Merge remote-tracking branch 'origin/unified-build' into unified-build

Biont committed:
:recycle: refactor(build): Add conditional PHP steps based on composer.json detection

This change refactors the build workflow to conditionally execute PHP-related steps only when a composer.json file is present. It introduces a check for composer.json at the beginning of the workflow and uses this detection to control the execution of PHP setup, Composer dependency installation, version updates, and PHP-specific tools like the WordPress Translation Downloader and PHP-Scoper. This makes the workflow more flexible and suitable for both PHP-based projects (WordPress plugins/themes, PHP libraries) and frontend-only JavaScript libraries, automatically skipping PHP-related steps when composer.json is not found.

The workflow now properly distinguishes between:
- PHP-based projects that require Composer setup and dependency management
- Frontend-only projects that should skip all PHP-related steps

Biont committed:
:recycle: refactor(build): Do composer.json check earlier

Biont committed:
:recycle: refactor(build): Do composer.json check EVEN earlier

Biont committed:
:memo: docs(build-and-distribute): Fix linter errors

Biont committed:
Update .github/workflows/build-and-distribute.yml

Co-authored-by: Philipp Bammes <8144115+tyrann0us@users.noreply.github.com>
Signed-off-by: Moritz Meißelbach <arbelzapf@gmail.com>

Biont committed:
Update .github/workflows/build-and-distribute.yml

Co-authored-by: Philipp Bammes <8144115+tyrann0us@users.noreply.github.com>
Signed-off-by: Moritz Meißelbach <arbelzapf@gmail.com>

Biont committed:
:memo: docs(build-and-distribute): Remove warning about cancel-in-progress setting

Biont committed:
Merge remote-tracking branch 'origin/unified-build' into unified-build

tyrann0us reviewed (approved):
Thank you for continuously implementing all the changes! From my side, this is good to be merged. @dottxado, @Chrico, any last wishes? 😆
